plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/snapshot/LinstorVMSnapshotStrategy.java

```diff
@@ -0,0 +1,347 @@
+        VMSnapshotVO vmSnapshotVO = (VMSnapshotVO) vmSnapshot;
+        if (vmSnapshotVO.getType() != VMSnapshot.Type.Disk) {
+            return StrategyPriority.CANT_HANDLE;
+        }
+        return allVolumesOnLinstor(vmSnapshot.getVmId());
+    }
+
+    private StrategyPriority allVolumesOnLinstor(Long vmId) {
+        List<VolumeObjectTO> volumeTOs = _vmSnapshotHelper.getVolumeTOList(vmId);
+        if (volumeTOs == null || volumeTOs.isEmpty()) {
+            return StrategyPriority.CANT_HANDLE;
+        }
+        for (VolumeObjectTO volumeTO : volumeTOs) {
+            Long poolId = volumeTO.getPoolId();
+            StoragePoolVO pool = _storagePoolDao.findById(poolId);
+            if (!pool.getStorageProviderName().equals(LinstorUtil.PROVIDER_NAME)) {
+                return StrategyPriority.CANT_HANDLE;
+            }
+        }
+        return StrategyPriority.HIGHEST;
+    }
+
+    private String linstorDeleteSnapshot(final DevelopersApi api, final String rscName, final String snapshotName) {
+        String resultMsg = null;
+        try {
+            ApiCallRcList answers = api.resourceSnapshotDelete(rscName, snapshotName, Collections.emptyList());
+            if (answers.hasError()) {
+                resultMsg = LinstorUtil.getBestErrorMessage(answers);
+            }
+        } catch (ApiException apiEx) {
+            log.error("Linstor: ApiEx - " + apiEx.getBestMessage());
+            resultMsg = apiEx.getBestMessage();
+        }
+
+        return resultMsg;
+    }
+
+    @Override
+    public boolean deleteVMSnapshot(VMSnapshot vmSnapshot) {
+        UserVmVO userVm = _userVmDao.findById(vmSnapshot.getVmId());
+        VMSnapshotVO vmSnapshotVO = (VMSnapshotVO) vmSnapshot;
+        try {
+            _vmSnapshotHelper.vmSnapshotStateTransitTo(vmSnapshot, VMSnapshot.Event.ExpungeRequested);
+        } catch (NoTransitionException e) {
+            log.debug("Failed to change vm snapshot state with event ExpungeRequested");
+            throw new CloudRuntimeException(
+                    "Failed to change vm snapshot state with event ExpungeRequested: " + e.getMessage());
+        }
+
+        List<VolumeObjectTO> volumeTOs = _vmSnapshotHelper.getVolumeTOList(vmSnapshot.getVmId());
+        final StoragePoolVO storagePool = _storagePoolDao.findById(volumeTOs.get(0).getPoolId());
+        final DevelopersApi api = LinstorUtil.getLinstorAPI(storagePool.getHostAddress());
+
+        final String snapshotName = vmSnapshotVO.getName();
+        for (VolumeObjectTO volumeObjectTO : volumeTOs) {
+            final String rscName = LinstorUtil.RSC_PREFIX + volumeObjectTO.getUuid();
+            String err = linstorDeleteSnapshot(api, rscName, snapshotName);
+
+            if (err != null)
+            {
+                throw new CloudRuntimeException(
+                    String.format("Unable to delete Linstor resource %s snapshot %s: %s", rscName, snapshotName, err));
```

Comment:
No, as deleting isn't time critical or is required to be bundled.
But it makes propably more sense to keep trying to delete all snapshots, and then at the end fail and report which failed?
Just ignoring seems also to be wrong, otherwise an admin would never notice orphaned vm-snapshots.

Reply:
@rp- maybe, better to continue delete other snapshots as well, and report the failed ones.